docs/adr/0009-infrastructure-as-code-sharing-approach.md

```diff
@@ -0,0 +1,94 @@
+# infrastructure-as-code-sharing-approach
+
+## Status
+
+Proposed
+
+## Context
+
+* The current version of Marain uses ARM templates to automate the infrastructure deployment for each Marain service.
+* The deployment templates resides alongside the code for each service
+* There is significant commonality/duplication of ARM templates between services
+* We wish to migrate to using Bicep to define the Marain infrastructure deployments
+* A desire to minimise duplication and utilise shared Bicep modules
+* Ideally, the overall Marain deployment process should be as low-friction as the current "'git clone' & deploy" approach
+
+Given this context and goals, the challenge is how to combine the use of shared Bicep modules (to de-duplicate the infrastructure-as-code) whilst still having a low friction means of deploying the Marain stack.  The absence of a defacto public Bicep module repository that can host the Marain-specific modules means there isn't an obvious way to share these modules without additional work.
+
+***NOTE**: A more general, but related, Bicep reuse [ADR](https://github.com/endjin/Endjin.RecommendedPractices.Bicep/blob/main/docs/adr/0002-sharing-bicep-modules.md) was written before it became clear that the public Bicep registry would not be 'open to all' in the same way that nuget.org or Docker Hub are.*
+
+In the ideal world the Marain deployment would use Bicep modules in the [Endjin.RecommendedPractices.Bicep repository](https://github.com/endjin/Endjin.RecommendedPractices.Bicep) so as to avoid duplication as well as a set of Marain-specific modules that are re-usable across different Marain services.  This means that the envisioned deployment process would need to interact with 3 sets of Bicep modules:
+
+1. Shared/Common - non-Marain specific functionality 
+1. Marain shared - functionality common to multiple Marain services
+1. Marain service-specific - functionality tailored to an individual Marain service
+
+This ADR considers the following options:
+
+1. Hosting a public Bicep registry
+1. Supporting internal Bicep registries
+1. Shipping pre-compiled ARM templates
+
+## Options
+
+### Hosting a public Bicep registry
+
+This option involves us hosting an Azure Container Registry that we make publicly available and publishing the Marain Bicep modules to it as part of the overall Marain release process.
+
+#### Advantages
+* Simplest and lowest friction option for the Marain consumer.
+* Easy to integrate with our existing release process.
+* Some flexibility for where the source for the shared Bicep modules is stored, as they won't be primarily consumed as 'source' artefacts.
```

Comment:
>they won't be primarily consumed as 'source' artefacts.

I don't understand what this means

Reply:
I've pushed-up some changes to, hopefully, better describe this